src/main/kotlin/org/rust/cargo/toolchain/impl/CargoMetadata.kt

```diff
@@ -220,11 +245,30 @@ object CargoMetadata {
     fun clean(project: Project): CargoWorkspaceData {
         val fs = LocalFileSystem.getInstance()
         val members = project.workspace_members
-            ?: error("No `members` key in the `cargo metadata` output.\n" +
-            "Your version of Cargo is no longer supported, please upgrade Cargo.")
+            ?: error("No `workspace_members` key in the `cargo metadata` output.\n" +
+                "Your version of Cargo is no longer supported, please upgrade Cargo.")
+
         return CargoWorkspaceData(
-            project.packages.mapNotNull { it.clean(fs, it.id in members) },
-            project.resolve.nodes.associate { (id, dependencies, deps) ->
+            project.packages.mapNotNull {
+                // resolve contains all enabled features for each package
+                val resolveNode = project.resolve.nodes.find { node -> node.id == it.id }
+                if (resolveNode == null) {
+                    LOG.error("Could not find package with `id` '${it.id}' in `resolve` section of the `cargo metadata` output.")
+                }
+
+                val enabledFeatures = resolveNode?.features
+                val features = it.features.keys.map { feature ->
+                    val state = when {
+                        enabledFeatures == null -> CargoWorkspace.FeatureState.Unknown
+                        enabledFeatures.contains(feature) -> CargoWorkspace.FeatureState.Enabled
```

Comment:
I'm not sure that it's a real issue but the size of `resolveNode.features` can be pretty big (for example, ~300 features in winapi 0.3). And in such cases `enabledFeatures.contains(feature)` for each feature can be relatively long if `enabledFeatures` is stored in list